node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -69,16 +69,26 @@ final case class TransactionAPIImpl[F[_]: Concurrent](
   ) = findTransactions(report).map(t => TransactionInfo(t, txCtor(d.deployInfo.get.sig)))
 
   def getTransaction(blockHash: Blake2b256Hash): F[List[TransactionInfo]] =
-    blockReportAPI.blockReport(blockHash.toByteString, forceReplay = false).map { res =>
+    blockReportAPI.blockReport(blockHash.toByteString, forceReplay = false).flatMap { res =>
       res
-        .map { b =>
-          val userTransactions = b.deploys.flatMap(d => {
-            d.report
-              .zip(Seq(PreCharge, UserDeploy, Refund))
-              .flatMap {
+        .traverse { b =>
+          val userTransactions = b.deploys.toList.flatTraverse(d => {
+            // FIXME https://github.com/rchain/rchain/issues/3150#issuecomment-701123669
+            val zippedReport =
+              if (d.report.length == 2)
+                d.report.zip(Seq(PreCharge, Refund)).pure
+              else if (d.report.length == 3) d.report.zip(Seq(PreCharge, UserDeploy, Refund)).pure
+              else
+                Concurrent[F].raiseError[Seq[(SingleReport, String => TransactionType)]](
+                  new Exception(
+                    s"It is not possible that user report ${d} amount is not equal to 3 or 2"
+                  )
+                )
+            zippedReport
+              .map(_.flatMap {
```

Comment:
Maybe it is not written well in fp style. If you got better writing ,please let me know. @tgrospic